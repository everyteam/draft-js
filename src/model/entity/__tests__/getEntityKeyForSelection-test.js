/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails isaac, oncall+ui_infra
 */

jest.disableAutomock();

var Immutable = require('immutable');
var getEntityKeyForSelection = require('getEntityKeyForSelection');
var getSampleStateForTesting = require('getSampleStateForTesting');
var matchers = require('jest-immutable-matchers');

var {contentState, selectionState} = getSampleStateForTesting();

selectionState = selectionState.merge({
  anchorKey: 'b',
  focusKey: 'b',
});

function setEntityMutability(mutability) {
  contentState.getEntityMap().get = () => ({
    getMutability: () => mutability,
  });
}
describe('getEntityKeyForSelection', () => {
  beforeEach(function() {
    jest.addMatchers(matchers);
  });

  describe('collapsed selection', () => {
    var collapsed = selectionState.merge({
      anchorOffset: 2,
      focusOffset: 2,
    });

    it('must return null at start of block', () => {
      var key = getEntityKeyForSelection(contentState, selectionState);
      expect(key).toBe(null);
    });

    it('must return key if mutable', () => {
      setEntityMutability('MUTABLE');
      var key = getEntityKeyForSelection(contentState, collapsed);
      expect(key).toEqualImmutable(Immutable.OrderedSet.of('123'));
    });

    it('must not return key if immutable', () => {
      setEntityMutability('IMMUTABLE');
      var key = getEntityKeyForSelection(contentState, collapsed);
      expect(key).toEqualImmutable(Immutable.OrderedSet());
    });

    it('must not return key if segmented', () => {
      setEntityMutability('SEGMENTED');
      var key = getEntityKeyForSelection(contentState, collapsed);
      expect(key).toEqualImmutable(Immutable.OrderedSet());
    });
  });

  describe('non-collapsed selection', () => {
    var nonCollapsed = selectionState.merge({
      anchorOffset: 2,
      focusKey: 'c',
      focusOffset: 2,
    });

    it('must return null if start is at end of block', () => {
      var startsAtEnd = nonCollapsed.merge({
        anchorOffset: contentState.getBlockForKey('b').getLength(),
      });
      var key = getEntityKeyForSelection(contentState, startsAtEnd);
      expect(key).toBe(Immutable.OrderedSet());
    });

    it('must return key if mutable', () => {
      setEntityMutability('MUTABLE');
      var key = getEntityKeyForSelection(contentState, nonCollapsed);
      expect(key).toEqualImmutable(Immutable.OrderedSet.of('123'));
    });

    it('must not return key if immutable', () => {
      setEntityMutability('IMMUTABLE');
      var key = getEntityKeyForSelection(contentState, nonCollapsed);
      expect(key).toBe(Immutable.OrderedSet());
    });

    it('must not return key if segmented', () => {
      setEntityMutability('SEGMENTED');
      var key = getEntityKeyForSelection(contentState, nonCollapsed);
      expect(key).toBe(Immutable.OrderedSet());
    });
  });
});
