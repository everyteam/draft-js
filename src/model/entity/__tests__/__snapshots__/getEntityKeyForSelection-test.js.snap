// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must not return key if immutable 1`] = `null`;

exports[`must not return key if immutable with collapsed selection 1`] = `null`;

exports[`must not return key if mutable with collapsed selection at end of an entity 1`] = `null`;

exports[`must not return key if mutable_interior and on edge 1`] = `null`;

exports[`must not return key if mutable_interior and on edge 2`] = `null`;

exports[`must not return key if mutable_interior and on edge 3`] = `null`;

exports[`must not return key if mutable_interior and on edge 4`] = `null`;

exports[`must not return key if segmented 1`] = `null`;

exports[`must not return key if segmented with collapsed selection 1`] = `null`;

exports[`must return key if mutable 1`] = `
Immutable.OrderedSet [
  "123",
]
`;

exports[`must return key if mutable with collapsed selection 1`] = `
Immutable.OrderedSet [
  "123",
]
`;

exports[`must return key if mutable_interior 1`] = `
Immutable.OrderedSet [
  "123",
]
`;

exports[`must return key if mutable_interior 2`] = `
Immutable.OrderedSet [
  "123",
]
`;

exports[`must return key if the start is at the end of a block and the entity is mutable_interior 1`] = `
Immutable.OrderedSet [
  "123",
]
`;

exports[`must return mutable_interior key across previous and next blocks 1`] = `
Immutable.OrderedSet [
  "123",
]
`;

exports[`must return mutable_interior key across previous and next blocks 2`] = `
Immutable.OrderedSet [
  "123",
]
`;

exports[`must return null at start of block with collapsed selection 1`] = `null`;

exports[`must return null if start is at end of block 1`] = `null`;
